commit 1792225c7006f6598931b3c4f2dbb80825c23397
Author: Sudeep Holla <sudeep.holla@arm.com>
Date:   Fri Jan 6 12:34:30 2017 +0000

    clk: scpi: don't add cpufreq device if the scpi dvfs node is disabled
    
    commit 67bcc2c5f1da8c5bb58e72354274ea5c59a3950a upstream.
    
    Currently we add the virtual cpufreq device unconditionally even when
    the SCPI DVFS clock provider node is disabled. This will cause cpufreq
    driver to throw errors when it gets initailised on boot/modprobe and
    also when the CPUs are hot-plugged back in.
    
    This patch fixes the issue by adding the virtual cpufreq device only if
    the SCPI DVFS clock provider is available and registered.
    
    Fixes: 9490f01e2471 ("clk: scpi: add support for cpufreq virtual device")
    Reported-by: Michał Zegan <webczat_200@poczta.onet.pl>
    Cc: Neil Armstrong <narmstrong@baylibre.com>
    Signed-off-by: Sudeep Holla <sudeep.holla@arm.com>
    Tested-by: Michał Zegan <webczat_200@poczta.onet.pl>
    Signed-off-by: Stephen Boyd <sboyd@codeaurora.org>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

diff --git a/drivers/clk/clk-scpi.c b/drivers/clk/clk-scpi.c
index 2a3e9d8..96d3717 100644
--- a/drivers/clk/clk-scpi.c
+++ b/drivers/clk/clk-scpi.c
@@ -290,13 +290,15 @@ static int scpi_clocks_probe(struct platform_device *pdev)
 			of_node_put(child);
 			return ret;
 		}
-	}
-	/* Add the virtual cpufreq device */
-	cpufreq_dev = platform_device_register_simple("scpi-cpufreq",
-						      -1, NULL, 0);
-	if (IS_ERR(cpufreq_dev))
-		pr_warn("unable to register cpufreq device");
 
+		if (match->data != &scpi_dvfs_ops)
+			continue;
+		/* Add the virtual cpufreq device if it's DVFS clock provider */
+		cpufreq_dev = platform_device_register_simple("scpi-cpufreq",
+							      -1, NULL, 0);
+		if (IS_ERR(cpufreq_dev))
+			pr_warn("unable to register cpufreq device");
+	}
 	return 0;
 }
 
